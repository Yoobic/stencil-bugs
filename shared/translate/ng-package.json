{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../node_modules/@shared/translate",
  "deleteDestPath": false,
  "lib": {
    "entryFile": "src/public_api.ts",
    "umdModuleIds": {
      "lodash-es": "lodash-es",
      "moment": "moment",
      "rxjs": "rxjs",
      "@shared/common": "@shared/common",
      "@ngx-translate/core": "@ngx-translate/core",
      "@shared/interfaces": "@shared/interfaces"
    }
  }
}