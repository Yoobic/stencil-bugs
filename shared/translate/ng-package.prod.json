{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
    "dest": "../../node_modules/@shared/translate",
  "lib": {
    "entryFile": "src/public_api.ts",
    "styleIncludePaths": [
      "../../../shared/styles"
    ]
  }
}