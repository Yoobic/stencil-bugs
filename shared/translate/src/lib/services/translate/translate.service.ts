import { Injectable } from '@angular/core';
import { CoreConfig, Persistent } from '@shared/common';
import { TranslateService, MissingTranslationHandler, MissingTranslationHandlerParams } from '@ngx-translate/core';
import { Observable, Subject } from 'rxjs';
export const AVAILABLE_LANGUAGES = [
    'en',
    'us',
    'fr',
    'de',
    'es',
    'nl',
    'pl',
    'it',
    'ru',
    'zhs',
    'zht',
    'pt',
    'kr',
    'ja',
    'ua',
    'he',
    'ar',
    'cz',
    'th',
    'tr',
    'bg',
    'el',
    'sl',
    'sk'
];

import { moment } from '@shared/interfaces';

@Injectable()
export class Translate {
    public static availablesLanguage: Array<string> = AVAILABLE_LANGUAGES;

    public static availablesLanguageAll: Array<string> = ['all'].concat(Translate.availablesLanguage);

    public static currentLanguage: string = '';

    public static loadExtraLanguage: (language) => Promise<any> = null;

    @Persistent('defaultLanguage') private defaultLanguage;

    private _languageChange = new Subject<string>();

    public static checkLanguage(language: string) {
        if (language === 'zh-Hant-HK' || language === 'zht') {
            language = 'zht';
        } else if (language === 'zh-Hans-GB' || language.startsWith('zh')) {
            language = 'zhs';
        } else {
            language = language.substring(0, 2);
        }
        if (Translate.availablesLanguage.indexOf(language) < 0) {
            language = 'en';
        }
        return language;
    }

    // public static loadFile(language: string) {
    //     language = Translate.checkLanguage(language);
    //     return System.import(/* webpackChunkName: "translate" */'../json/' + (language || 'en') + '.json').then(translations => {
    //         Translate[language] = translations;
    //         if (Translate.loadExtraLanguage) {
    //             return Translate.loadExtraLanguage(language);
    //         } else {
    //             return Promise.resolve({});
    //         }
    //     }).then(extras => {
    //         Translate[language] = Object.assign(Translate[language], extras || {});
    //         return Translate[language];
    //     });
    // }

    constructor(private ngTranslate: TranslateService, private coreConfig: CoreConfig) { }

    init(): Promise<void> {
        let browserLanguage = 'en';
        //this.ngTranslate.addLangs(AVAILABLE_LANGUAGES);
        if (!this.coreConfig.isUniversal()) {
            browserLanguage = (<any>window).navigator.userLanguage || window.navigator.language || 'en';
        }
        // if (this.coreConfig.isCordova() && !this.defaultLanguage) {
        //     return this.globalization.getPreferredLanguage().then(
        //         res => this.setLanguage(res.value),
        //         err => this.setLanguage(browserLanguage)
        //     );
        // } else {
        this.defaultLanguage = this.defaultLanguage || browserLanguage;
        return this.setLanguage(this.defaultLanguage);
        //        }
    }

    get languageChange$(): Observable<string> {
        return this._languageChange.asObservable();
    }

    setLanguage(language: string): Promise<any> {
        return new Promise((resolve, reject) => {
            if (language) {
                language = Translate.checkLanguage(language);
                this.defaultLanguage = language;
                this.ngTranslate.setDefaultLang(this.defaultLanguage);
                moment.locale(this.getLanguageFull(language, '-'));
                this._languageChange.next(language);
                Translate.currentLanguage = language;
                if (typeof window !== 'undefined' && window.document) {
                    if (this.isRtl(language)) {
                        window.document.dir = 'rtl';
                    } else {
                        window.document.dir = 'ltr';
                    }
                }
                this.ngTranslate.use(language).subscribe(() => {
                    resolve(language);
                });
            } else {
                resolve(null);
            }
        });
    }

    addDynamicLanguage(language: string, translations: Array<any>, isAdmin: boolean) {
        let newTranslations = (<any>this.ngTranslate).translations[language] || {};
        translations.forEach(d => {
            if (d && d.key && (!isAdmin || this.ngTranslate.instant(d.key) === d.key)) {
                newTranslations[d.key] = d[language];
            }
        });
        this.ngTranslate.setTranslation(language, newTranslations, true);
    }

    getCurrentLanguage() {
        return this.defaultLanguage;
    }

    getCurrentAngularLocale() {
        switch (this.defaultLanguage) {
            case 'en':
                return 'en-GB';
            case 'us':
                return 'en-US';
            case 'zhs':
                return 'zh-Hans';
            case 'zht':
                return 'zh-Hant';
            case 'fr':
            case 'de':
            case 'es':
            case 'nl':
            case 'pl':
            case 'it':
            case 'ru':
            case 'pt':
            case 'ja':
            case 'he':
            case 'ar':
            case 'th':
            case 'tr':
            case 'bg':
            case 'gr':
            case 'sl':
            case 'sk':
                return this.defaultLanguage;
            default:
                return 'en-US';
        }
    }

    getCurrentLanguageFull(separator = '_') {
        let l = this.getCurrentLanguage();
        return this.getLanguageFull(l, separator);
    }

    get(key: string, params?: Object): string {
        try {
            return this.ngTranslate.instant(key, params);
        } catch (err) {
            return key;
        }
    }

    getIcon(language: string) {
        switch (language) {
            case 'en':
                return 'flag-icon flag-icon-gb';
            case 'zhs':
                return 'flag-icon flag-icon-cn';
            case 'zht':
                return 'flag-icon flag-icon-hk';
            case 'ja':
                return 'flag-icon flag-icon-jp';
            case 'he':
                return 'flag-icon flag-icon-il';
            case 'ar':
                return 'flag-icon flag-icon-ae';
            case 'el':
                return 'flag-icon flag-icon-gr';
            case 'sl':
                return 'flag-icon flag-icon-si';
            default:
                return 'flag-icon flag-icon-' + language;
        }
    }

    getAll(language: string) {
        return Translate[language];
    }

    // loadAllFiles() {
    //     return Promise.all(Translate.availablesLanguage.map(l => Translate.loadFile(l)));
    // }

    polyglot(value: string) {
        if (value && value.replace && value.indexOf('%') >= 0) {
            return value.replace(/%(\w+)%/g, (match, key) => {
                return this.get(key);
            });
        }
        if (value && value.toUpperCase && value.toUpperCase() === value) {
            return this.get(value);
        }
        return value;
    }

    isCurrentLanguageRtl() {
        return this.isRtl(this.getCurrentLanguage());
    }

    isCurrentLanguageChinese() {
        let l = this.getCurrentLanguage();
        return ['zhs'].indexOf(l) >= 0;
    }

    isRtl(language: string) {
        if (language === 'he' || language === 'ar') {
            return true;
        }
        return false;
    }

    getLanguageFull(language: string, separator = '_') {
        switch (language) {
            case 'us':
                return 'en' + separator + 'US';
            case 'en':
                return 'en' + separator + 'GB';
            case 'ua':
                return 'uk' + separator + 'UA';
            case 'zhs':
                return 'zh' + separator + 'CN';
            case 'zht':
                return 'zh' + separator + 'TW';
            case 'ae':
                return 'ar' + separator + 'AE';
            default:
                return language + separator + language.toUpperCase();
        }
    }
}

@Injectable()
export class LogMissingTranslationHandler implements MissingTranslationHandler {
    handle(params: MissingTranslationHandlerParams) {
        return params.key;
    }
}

// @Injectable()
// export class CustomTranslateLoader implements TranslateLoader {
//     getLanguage(language: string) {
//         return Translate.loadFile(language);
//     }

//     getTranslation(lang: string): Observable<any> {
//         return new Observable<any>((observer: Observer<any>) => {
//             this.getLanguage(lang).then(val => {
//                 observer.next(val);
//                 observer.complete();
//             });
//         });
//     }
// }
