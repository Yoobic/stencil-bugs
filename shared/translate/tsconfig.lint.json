{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "../../out-tsc/spec",
    "module": "esnext",
    "types": []
  },
  "exclude": [
    "src/test.ts",
    "**/*.spec.ts"
  ]
}
