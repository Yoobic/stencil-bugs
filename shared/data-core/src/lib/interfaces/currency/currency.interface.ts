import { Model } from '../../decorators/model/model.decorator';
import { Searchable } from '../../decorators/searchable/searchable.decorator';
import { Editable } from '../../decorators/editable/editable.decorator';
import { FormFieldType, ICurrency } from '@shared/interfaces';
import { Translate } from '@shared/translate';

@Model({
    className: 'Currency',
    collectionName: 'currencies',
    fields: ['*'],
    include: []
})

export class Currency extends ICurrency {

    @Editable('Currency', { required: true, type: FormFieldType.autocomplete, title: 'LANGUAGE', translate: true, values: Translate.availablesLanguage, clearable: false })
    @Searchable('Currency') currency: string;

    @Editable('Currency', { required: true, type: FormFieldType.text })
    @Searchable('Currency') symbol: string;

    @Editable('Currency', { required: true, type: FormFieldType.number, title: 'RATE1' })
    @Searchable('Currency') rate: number;

}
