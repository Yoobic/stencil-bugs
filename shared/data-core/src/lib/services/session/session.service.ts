import { Injectable } from '@angular/core';
//import { User } from '../../interfaces/user/user.interface';
import { Currency } from '../../interfaces/currency/currency.interface';
import { Location } from '../../interfaces/location/location.interface';
import { MissionDescription } from '../../interfaces/mission-description/mission-description.interface';
import { Persistent, LocalForageService } from '@shared/common';

@Injectable()
export class Session {

    //@Persistent('session.token')
    public token: string;
    //@Persistent('session.user')
    public user: any;
    //@Persistent('session.userId')
    public userId: string;
    //@Persistent('session.currencies')
    public currencies: Array<Currency>;
    //@Persistent('session.groups')
    public groups: Array<string>;
    //@Persistent('session.roles')
    public roles: Array<string>;

    public tenantName?: string;
    public tenantRef?: string;

    @Persistent('login.hideWalkthrough') public hideWalkthrough: boolean;

    public openedChannels: Array<string> = [];

    public selectedMissionDescription: MissionDescription = null;
    public selectedLocation: Location = null;

    constructor(protected localForage: LocalForageService) { }

    clear(clearUser = true) {
        if (clearUser) {
            this.token = '';
            this.user = null;
            this.userId = null;
            this.tenantName = null;
            this.tenantRef = null;
            this.currencies = [];
            this.groups = [];
            this.roles = [];
            this.hideWalkthrough = false;
        }
        return Promise.resolve();
    }
}
