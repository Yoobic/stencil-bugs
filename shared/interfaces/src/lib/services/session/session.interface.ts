import { ICurrency } from '../../entities/currency/currency.interface';

export interface ISessionService {
    token: string;
    user: any;
    userId: string;
    currencies: Array<ICurrency>;
    groups: Array<string>;
    roles: Array<string>;
    tenantName?: string;
    tenantRef?: string;
}