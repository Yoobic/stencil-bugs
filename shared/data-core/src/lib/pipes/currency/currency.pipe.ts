import { Pipe, PipeTransform } from '@angular/core';
import { DecimalPipe } from '@angular/common';
import { Session } from '../../services/session/session.service';
import { Translate } from '@shared/translate';
import { filter } from 'lodash-es';

@Pipe({ name: 'currency' })
export class CurrencyPipe implements PipeTransform {

    private translate: Translate;
    private decimalPipe: DecimalPipe;

    constructor(translate: Translate, private session: Session) {
        this.translate = translate;
        this.decimalPipe = new DecimalPipe(this.translate.getCurrentAngularLocale());
    }

    transform(value: number, ...args: string[]): any {
        let precision = args && args.length > 0 ? args[0] : 2;

        if (!value && value !== 0) {
            return '';
        }
        if (Math.abs(value) >= 1 || value === 0) {
            precision = precision || 0;
        } else {
            precision = precision || 1;
        }
        let symbol = this.translate.get('CURRENCY');
        let rates = filter(this.session.currencies, c => c.symbol === symbol);

        let rate = 1;
        if (rates.length > 0) {
            rate = rates[0].rate;
        }

        let displayValue = this.decimalPipe.transform(value * rate, '1.0-' + precision.toString());

        if (symbol === 'Â£') {
            return symbol + ' ' + displayValue;
        } else {
            return displayValue + ' ' + symbol;
        }
    }
}
