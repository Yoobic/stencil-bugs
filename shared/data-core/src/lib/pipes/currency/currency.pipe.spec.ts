import { CurrencyPipe } from './currency.pipe';
import { Session } from '../../services/session/session.service';
import { Translate } from '@shared/translate';

describe('pipe: CurrencyPipe', () => {

    const CURRENCIES = [
        {
            'rate' : 1.0,
            'currency' : 'fr',
            'symbol' : '€'
        }, {
            'rate' : 0.85,
            'currency' : 'en',
            'symbol' : '£'
        }, {
            'rate' : 1.05,
            'currency' : 'us',
            'symbol' : '$'
        }, {
            'rate' : 1.0,
            'currency' : 'es',
            'symbol' : '€'
        }, {
            'rate' : 4.3,
            'currency' : 'pl',
            'symbol' : 'zł'
        },  {
            'rate' : 7.2,
            'currency' : 'zht',
            'symbol' : '¥'
        }, {
            'rate' : 3.8,
            'currency' : 'he',
            'symbol' : '₪'
    }].map (i =>  Object.assign(i, {_id: null, _acl: null}));

    let mockTranslate: Translate;
    let mockSession: Session;
    let currencyPipe: CurrencyPipe;

    let pi: number = 3.14156;
    let little: number = -0.12345;
    let rules = [
        {currency: '$', locale: 'en-US'},
        {currency: '€', locale: 'fr-FR'},
        {currency: '£', locale: 'en-GB'},
        {currency: '₪', locale: 'he-IL'}
    ];

    beforeEach( () => {
        mockTranslate = new Translate(null, null, null);
        mockSession = new Session(null);
        mockSession.currencies = CURRENCIES;
        currencyPipe = new CurrencyPipe(mockTranslate, mockSession);
    });

    rules.forEach ( i => {
        it(`show currency for  ${i.currency}, locale: ${i.locale}`, () => {
            mockTranslate.get = jest.fn(() => i.currency);
            mockTranslate.getCurrentAngularLocale = jest.fn(() => i.locale);
            expect(currencyPipe.transform(pi)).toMatchSnapshot();
            expect(currencyPipe.transform(pi, '1')).toMatchSnapshot();

            expect(currencyPipe.transform(little)).toMatchSnapshot();

            expect(currencyPipe.transform(little, '1')).toMatchSnapshot();

        });

    });

    it ('should return empty if no value', () => {
        mockTranslate.get = jest.fn(() => '€');
        mockTranslate.getCurrentAngularLocale = jest.fn(() => 'en-US');
        expect(currencyPipe.transform(undefined)).toEqual('');

    });
});