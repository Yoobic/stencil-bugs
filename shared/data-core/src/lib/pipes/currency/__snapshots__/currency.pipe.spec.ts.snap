// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipe: CurrencyPipe show currency for  $, locale: en-US 1`] = `"3.3 $"`;

exports[`pipe: CurrencyPipe show currency for  $, locale: en-US 2`] = `"3.3 $"`;

exports[`pipe: CurrencyPipe show currency for  $, locale: en-US 3`] = `"-0.13 $"`;

exports[`pipe: CurrencyPipe show currency for  $, locale: en-US 4`] = `"-0.1 $"`;

exports[`pipe: CurrencyPipe show currency for  £, locale: en-GB 1`] = `"£ 2.67"`;

exports[`pipe: CurrencyPipe show currency for  £, locale: en-GB 2`] = `"£ 2.7"`;

exports[`pipe: CurrencyPipe show currency for  £, locale: en-GB 3`] = `"£ -0.1"`;

exports[`pipe: CurrencyPipe show currency for  £, locale: en-GB 4`] = `"£ -0.1"`;

exports[`pipe: CurrencyPipe show currency for  ₪, locale: he-IL 1`] = `"11.94 ₪"`;

exports[`pipe: CurrencyPipe show currency for  ₪, locale: he-IL 2`] = `"11.9 ₪"`;

exports[`pipe: CurrencyPipe show currency for  ₪, locale: he-IL 3`] = `"-0.47 ₪"`;

exports[`pipe: CurrencyPipe show currency for  ₪, locale: he-IL 4`] = `"-0.5 ₪"`;

exports[`pipe: CurrencyPipe show currency for  €, locale: fr-FR 1`] = `"3.14 €"`;

exports[`pipe: CurrencyPipe show currency for  €, locale: fr-FR 2`] = `"3.1 €"`;

exports[`pipe: CurrencyPipe show currency for  €, locale: fr-FR 3`] = `"-0.12 €"`;

exports[`pipe: CurrencyPipe show currency for  €, locale: fr-FR 4`] = `"-0.1 €"`;
